### PR TITLE
Implemented shell flag

### PR DESCRIPTION
**Pull request by SuperMaZingCoder**
# Overview
This flag allows launches a shell-like environment for running code via the Piston API. This is a WIP pr as of now, as it breaks the "flagless" piston-cli command from my testing. I also need to add newlines between the compilation and prompts for args and stdin as well as documenting that `esc + enter` is requred to send the code to the API. 

Other than that, the shell command is working perfectly!

# Screenshots
image

### CODE REVIEW

**Review by Shivansh-007 — COMMENTED** [review]
Nicely done! But in the python REPL we can't really give it system arguments, so I don't we need to give it here. Secondly I think the stdin arguments should be asked after the code is give.
And last, the language should be specified while running the command that is:
```bash
$ piston --shell python
```

**Comment by SuperMaZingCoder** [comment]
Thanks! I'm not sure about dropping system arguments to be better lined up with the python REPL; already, data is not stored between "REPL code runs" and I think that makes it a tool to repeatedly run the piston-cli. If that is the goal - to write all code once and test it out - I think that system arguments should be a provided option. If the distinction I made is a bit confusing or you still think we should drop system arguments, let me know!

As for putting stdin *after* the code prompt, I'm okay with this, though, I think we should change that in the "flagless" execution too if we do so.

**Comment by SuperMaZingCoder** [comment]
After some more testing, it seems like it doesn't break the other flags, but flagless piston is strangely asking for the language twice.

**Comment by Shivansh-007** [comment]
> Thanks! I'm not sure about dropping system arguments to be better lined up with the python REPL; already, data is not stored between "REPL code runs" and I think that makes it a tool to repeatedly run the piston-cli. If that is the goal - to write all code once and test it out - I think that system arguments should be a provided option. If the distinction I made is a bit confusing or you still think we should drop system arguments, let me know!
> 
> As for putting stdin _after_ the code prompt, I'm okay with this, though, I think we should change that in the "flagless" execution too if we do so.

Ok agree with the first point, and yeah you can put the stdin after the code prompt and maybe even the sys arguments (not sure about this).

File: piston/commands/from_shell.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        self.console = Console()
+        self.themes = list(get_all_styles()) + schemes
+        self.style = None
+        self.language = None
+        self.prompt_session = None
+
+    def set_prompt_session(self) -> None:
+        """Set the prompt session to use for input."""
+        self.prompt_session = PromptSession(
+            ">>> ",
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:35]: This should be a constant.

File: piston/commands/from_shell.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                self.style = scheme_dict[theme]()
+        else:
+            self.style = sfpc(get_style_by_name("solarized-dark"))
+
+    def set_language(self) -> None:
+        """Prompt the user for the programming language, close program if language not supported."""
+        language = self.console.input("[green]Enter language:[/green] ").lower()
+
+        if language not in languages_:
+            self.console.print("[bold red]Language is not supported![/bold red]")
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:57]: We are using the style argument as decided. (Am not putting this for all such cases but yeah it is meant for all).

File: piston/commands/from_shell.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    def get_stdin(self) -> str:
+        """Prompt the user for the standard input."""
+        stdin = self.console.input(
+            "[green]Enter your stdin arguments:[/green] "
+        ).lower()
+        return stdin
+
+    def query_piston(
+        self, code: str, args: List[str], style: Style, stdin: str
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:77]: Parameter `style` is not used

File: piston/commands/from_shell.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            ]
+            return "\n".join(result), self.language
+
+    def run_shell(self, theme: str) -> None:
+        """Run the shell."""
+        self.set_language()
+        self.set_prompt_session()
+        self.set_style(theme)
+        output = ""
+        while output != "exit":
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:122]: We can have a list of "Exit" words like `quit`, `bye`, `cya`.

File: piston/commands/from_shell.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            style=self.style,
+        )
+
+        if code.strip() == "exit":
+            return "exit", self.language
+
+        data = self.query_piston(code, args, self.style, stdin)
+
+        if len(data["output"]) == 0:
+            return "Your code ran without output.", self.language
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:108]: Add this as a constant as it is being used multiple times. Not in this file but yeah in few others.

File: piston/commands/from_shell.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        return [x for x in args.strip().split(",") if x]
+
+    def get_stdin(self) -> str:
+        """Prompt the user for the standard input."""
+        stdin = self.console.input(
+            "[green]Enter your stdin arguments:[/green] "
+        ).lower()
+        return stdin
+
+    def query_piston(
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:76]: We can make `query_piston` a utils function as it is being used in multiple files.

File: piston/commands/from_shell.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            "[green]Enter your args separated by comma:[/green] "
+        ).lower()
+        return [x for x in args.strip().split(",") if x]
+
+    def get_stdin(self) -> str:
+        """Prompt the user for the standard input."""
+        stdin = self.console.input(
+            "[green]Enter your stdin arguments:[/green] "
+        ).lower()
+        return stdin
```
Comment (Shivansh-007) [inline piston/commands/from_shell.py:74]: We can make this a utils function.

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]

**Comment by SuperMaZingCoder** [comment]
> Ok agree with the first point, and yeah you can put the stdin after the code prompt and maybe even the sys arguments (not sure about this).

Alright. We should document that the stdin/args comes *after* the code in this case. I think if we do this, we need to change it for `from_input` too, and possibly the others.

File: piston/commands/from_shell.py (hunk 1 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:35]: You mean the `">>> "`, right? I agree with that.

File: piston/commands/from_shell.py (hunk 2 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:57]: 👍 I'll change those, note that we'll have to do the same for the other execution methods in the future.

File: piston/commands/from_shell.py (hunk 3 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:77]: 👍

File: piston/commands/from_shell.py (hunk 4 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:122]: Sounds good, I'll add this to `constants.py`

File: piston/commands/from_shell.py (hunk 5 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:108]: Which part?

File: piston/commands/from_shell.py (hunk 1 above)
Comment (Shivansh-007) [reply piston/commands/from_shell.py:35]: Yeah!

File: piston/commands/from_shell.py (hunk 5 above)
Comment (Shivansh-007) [reply piston/commands/from_shell.py:108]: `Your code ran without output.`

Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:108]: I think we should save this for the time we format everything, and keep this PR just for shell execution.

File: piston/commands/from_shell.py (hunk 6 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:76]: I think we should save this for the time we format everything, and keep this PR just for shell execution.

File: piston/commands/from_shell.py (hunk 7 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:74]: I think we should save this for the time we format everything, and keep this PR just for shell execution.

File: piston/commands/from_shell.py (hunk 5 above)
Comment (Shivansh-007) [reply piston/commands/from_shell.py:108]: > I think we should save this for the time we format everything, and keep this PR just for shell execution.

Alright. No Problem.

File: piston/commands/from_shell.py (hunk 6 above)
Comment (Shivansh-007) [reply piston/commands/from_shell.py:76]: > I think we should save this for the time we format everything, and keep this PR just for shell execution.

Alright. No Problem.

File: piston/commands/from_shell.py (hunk 7 above)
Comment (Shivansh-007) [reply piston/commands/from_shell.py:74]: Alright. No Problem.

File: piston/commands/from_shell.py (hunk 1 above)
Comment (SuperMaZingCoder) [reply piston/commands/from_shell.py:35]: Sounds good! :D

**Comment by SuperMaZingCoder** [comment]
I've implemented what we discussed, though there is some strange behavior in that it does not display the compilation output until enter is clicked on the next prompt. Any ideas about what could be happening here?

**Comment by SuperMaZingCoder** [comment]
So, turns out that this was just an issue with Visual Studio's terminal, this should be good to merge at this point! Of course, if there is anything that needs to be fixed, let me know :D

**Review by Shivansh-007 — APPROVED** [review]

**Pull request merged**